mqtt/mqtt-broker/src/auth.rs

```diff
@@ -0,0 +1,85 @@
+#![allow(dead_code)] // TODO @dmolokanov remove when implemented
+use std::fmt;
+
+use failure::Fail;
+
+use crate::Error;
+
+#[derive(Clone, Debug, PartialEq)]
+pub enum AuthId {
+    Any,
+    Value(Identity),
+}
+
+pub type Identity = String;
+
+#[derive(Clone, Debug)]
+pub struct Certificate(Vec<u8>);
+
+pub struct Credentials {
+    username: Option<String>,
+    password: Option<String>,
+    certificate: Option<Certificate>,
+}
+
+impl Credentials {
+    pub fn new(
+        username: Option<String>,
+        password: Option<String>,
+        certificate: Option<Certificate>,
+    ) -> Self {
+        Self {
+            username,
+            password,
+            certificate,
+        }
+    }
+}
+
+pub trait Authenticator {
+    fn authenticate(&self, credentials: Credentials) -> Result<Option<AuthId>, Error>;
+}
```

Comment:
I wonder if the Option is needed around the AuthId. It is either authenticated (even if it becomes "any" which I see something like anonymous) or not, and if not, that is an error case. Because of that, I don't see a non-error coming back with None(). Is there any case I don't see?

Reply:
We want to distinguish cases
None - external component is not able to identify client, and
Err() - error occurred while making a call to an external component.